File: greedy/1403_minimum_subsequence_in_non_increasing_order/1403_minimum_subsequence_in_non_increasing_order.go
package main

import (
	"fmt"
	"sort"
)

func minSubsequence(nums []int) []int {
	var (
		i, sum, tmp int
		tmpSlice    []int
		length      = len(nums)
	)
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] > nums[j]
	})
	for i = 0; i < length; i++ {
		sum += nums[i]
	}
	for i = 0; i < length; i++ {
		tmp += nums[i]
		tmpSlice = append(tmpSlice, nums[i])
		if tmp > sum-tmp {
			return tmpSlice
		}
	}
	return tmpSlice
}

func main() {
	nums := []int{4, 3, 10, 9, 8}
	res := minSubsequence(nums)
	fmt.Println(res)
}

File: greedy/1403_minimum_subsequence_in_non_increasing_order/1403_minimum_subsequence_in_non_increasing_order.py
from typing import List

class Solution:
    def minSubsequence(self, nums: List[int]) -> List[int]:
        N = len(nums)

        total = sum(nums)

        nums.sort()

        ret = []
        t = 0
        while True:
            x = nums.pop()

            t += x
            ret.append(x)
            if t * 2 > total: return ret


nums = [4,3,10,9,8]
res = Solution().minSubsequence(nums)
print(res)